### DISCUSSION TITLE
Cannot debug a PE from a non-Windows platform

### DISCUSSION THREAD

**Discussion by 25pwn** [Vector35/binaryninja-api#5833] [Q&A]
I am running Fedora Rawhide.

In `Debug Adapter Settings`, `Adapter Type` is blank and clicking on it shows nothing. `binaryninja.debugger.debugadaptertype.DebugAdapterType.get_available_adapters(bv)` returns an empty `list`. I am unable to use remote debugging.

This used to work, but it stopped working some time ago. How do I fix this issue?

**Comment by psifertex** [comment]
Unfortunately Fedora is not a supported platform. That said, this sounds more like a corrupted download. Have you tried the usual uninstall, reinstall a fresh version, double check the hashes of your download? Try both latest stable and dev as well.

**Comment by xusheng6** [comment]
> I am running Fedora Rawhide.
> 
> 
> 
> In `Debug Adapter Settings`, `Adapter Type` is blank and clicking on it shows nothing. `binaryninja.debugger.debugadaptertype.DebugAdapterType.get_available_adapters(bv)` returns an empty `list`. I am unable to use remote debugging.
> 
> 
> 
> This used to work, but it stopped working some time ago. How do I fix this issue?

Could you please give more details on how you used it when it previous worked? Were you doing local debugging, or remote debugging? What is the thing you are debugging?

**Comment by 25pwn** [comment]
I'm pretty sure it's not a corruption, but I reinstalled just to be safe.

I was remote debugging windows and linux binaries.

I did a bit more testing and it seems the issue only happens on PE files. With ELF and macho I do get `["LLDB"]` in the dropdown and `binaryninja.debugger.debugadaptertype.DebugAdapterType.get_available_adapters(bv)`.

**Reply by xusheng6** [reply]
This sounds like https://github.com/Vector35/debugger/issues/527, which I should be able to fix soon. That said, I believe it should never have worked at all. Do you have a specific build that it works?

Also, how are you remote debugging Windows executable? That is a workflow I think we do not yet support

**Reply by 25pwn** [reply]
(I was trying to find time to debug this but I'll be busy for a while, so I'll say this for now.)

Sorry, I was conflating `Debug Adapter Settings` with `Debug Server Settings`. My workflow used to be `lldb-server p --server --listen 0.0.0.0:31337` on the target machine, and connecting to it using the `remote-*` platform.

On a Linux ELF binary it works fine: it lists all targets, and I can connect and debug just fine. On a Windows PE binary, It used to be possible to select a `Platform`(works on `4.0.5336`). Now it's blank, I cannot select a platform, and clicking `Accept` without  a platform gives me `Cannot connect to the debug server. Please check the connection configuration.`.

**Reply by xusheng6** [reply]
I believe that even if you can select something for a PE. it would not really work. Remote debugging a PE from Linux is not yet supported, and you can only debug a PE from another Windows machine. https://github.com/Vector35/debugger/issues/70

**Reply by 25pwn** [reply]
I'm talking about `lldb-server` remote debugging, not binary ninja's Windows debugger. This is definitely a regression, because it worked on `4.0.5336` but not on the latest build.

**Reply by xusheng6** [reply]
I might be confused, could you please state again what is the thing that you think was working but no longer works now? I thought you meant that you can do Windows PE remote debugging from Linux, which I think is impossible. But I could be wrong and we are talking about different things

**Reply by 25pwn** [reply]
I am aware that Binary Ninja's remote Windows debugger cannot be used from Linux. This is a regression in the handling of `lldb-server` remote debugging. They are entirely different debuggers.

My remote debugging workflow(Linux -> Windows) on Binary Ninja 4.0.5336 used to be:
1. `lldb-server p --server --listen 0.0.0.0:31337` on the target Windows machine
2. Activate `Connect to Debug Server` on the analysis Linux machine, and in the resulting `Debug Server Settings` menu, fill out the connection info and set `Platform` as `remote-windows`
3. Activate `Debug Adapter Settings` on the analysis Linux machine, and in the resulting `Debug Adapter Settings` menu, select `Adapter Type` as `LLDB`, `Input File` as the path to the target executable on the Windows machine(the executable must be copied to the Windows machine), `Executable Path` as the local path to the target executable, and `Working Directory` as the desired working directory or blank.
4. start debugging

This worked pretty well at the time(I had to restart Binary Ninja under some conditions, but it was perfectly usable, and that is an unrelated issue). I downgraded to that version to check before writing [this](https://github.com/Vector35/binaryninja-api/discussions/5833#discussioncomment-10506286).

As of the latest build, in step 2, the `Platform` dropdown is empty, and nothing can be selected. If I click `Accept` regardless, I get `Cannot connect to the debug server. Please check the connection configuration.`.

**Reply by xusheng6** [reply]
Hmm interesting. In my sense, you are actually doing Windows remote debugging from a Linux platform, despite using the lldb adapter, which I think is NOT supported and I remember I tested it and it did not work. I remember in LLDB 14, the lldb-server would just crash when you connect to it. Maybe things have changed in LLDB 16

**Reply by xusheng6** [reply]
we intentionally disabled the lldb adapter for PE files because it does not work. So it is intentional and not really a regression. But I will see if it actually works if I set up the things like you have done

**Reply by 25pwn** [reply]
I was using LLDB 18.1, and it worked just fine. Nothing seemed weird except for the occasional reliability issue, but that's a different issue.

**Reply by xusheng6** [reply]
Oh that is interesting. Maybe they added the support somewhere between 16.0 and 18.1. I will verify it it work!

Thanks for letting me know, btw! You could have saved me tons of time trying to have remote Windows debugging support from Linux/macOS through a different route

**Reply by 25pwn** [reply]
I have not tested any other LLDB versions, but the debug adapter configuration(step 3) is a bit counterintuitive, maybe during testing the tester got errors due to that and concluded that it does not work? Also may be the python version issue.

**Reply by xusheng6** [reply]
Just one more favor, where did you get the lldb 18.1 binaries? Did you build it yourself? Or you downloaded it somewhere

**Reply by 25pwn** [reply]
If you can't reproduce I'm happy to help set it up, I'm happy to contribute.

**Reply by xusheng6** [reply]
Well, I actually directly tested with the lldb/lldb-server binaries (not using the binja debugger), so I think I am correct that it does not work in LLDB 16.0. Probably they added it before 18.1 and I just did not know

**Reply by 25pwn** [reply]
I got the binaries from here: https://github.com/llvm/llvm-project/releases

**Reply by xusheng6** [reply]
Thanks in advance! I will test the setup and see if it works, and I will get back to you if there is any updates. Expect to hear from me soon!

**Reply by xusheng6** [reply]
> I got the binaries from here: https://github.com/llvm/llvm-project/releases

Sure, but I do not think that one has the lldb-server binaries in it, do they?

**Reply by 25pwn** [reply]
The `clang+llvm-*`s have `bin/lldb-server`. I think I recall having to install a non-latest python version at the time due to compatibility issues too, but I can't check the version because I'm not at my workstation atm. I can check in less than 24hrs if needed.

**Reply by xusheng6** [reply]
Yeah I just downloaded the ` clang+llvm-18.1.8-x86_64-pc-windows-msvc.tar.xz ` and there is a `bin/lldb-server.exe` in it. I think this does not exist in LLDB 16.0 build. This makes it more promising because maybe they included it because it now works!

**Reply by xusheng6** [reply]
Yeah it seems that LLDB 18.1 has made some improvement against 16.0. If I try to set the remote-debugging in LLDB 16.0, it actually crashes right after I try to connect to it:

```
PS C:\Users\xusheng\libclang\16.0.0\bin> .\lldb-server.exe p --server --listen 0.0.0.0:31337
Connection established.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.      Program arguments: C:\\Users\\xusheng\\libclang\\16.0.0\\bin\\lldb-server.exe p --server --listen 0.0.0.0:31337
1.      Program arguments: C:\\Users\\xusheng\\libclang\\16.0.0\\bin\\lldb-server.exe p --server --listen 0.0.0.0:31337
Exception Code: 0x80000003
#0 0x00007ff7a7108da0 (C:\Users\xusheng\libclang\16.0.0\bin\lldb-server.exe+0x68da0)
#1 0x00007ff7a710a5b7 (C:\Users\xusheng\libclang\16.0.0\bin\lldb-server.exe+0x6a5b7)
#2 0x00007ff7a710adb3 (C:\Users\xusheng\libclang\16.0.0\bin\lldb-server.exe+0x6adb3)
#3 0x00007ff7a811f67c (C:\Users\xusheng\libclang\16.0.0\bin\lldb-server.exe+0x107f67c)
#4 0x00007ffe592c257d (C:\Windows\System32\KERNEL32.DLL+0x1257d)
#5 0x00007ffe5a9caf28 (C:\Windows\SYSTEM32\ntdll.dll+0x5af28)
```

However, with LLDB 18.1, at least the connection can be established and the target can be executed properly:

```
(lldb) platform connect connect://127.0.0.1:31337
  Platform: remote-windows
    Triple: x86_64-pc-windows-msvc
OS Version: 6.2.0 (Windows NT 6.2.0)
  Hostname: DESKTOP-NTLKJTP
 Connected: yes
WorkingDir: C:\Users\xusheng\Downloads\clang+llvm-18.1.8-x86_64-pc-windows-msvc\bin
    Kernel: Windows NT 6.2.0
      Host: Windows 6.2.0
(lldb) r
(lldb) Process 10308 launched: 'C:\Users\xusheng\debugger_build\debugger\test\binaries\Windows-x86_64\helloworld.exe' (x86_64)
Process 10308 exited with status = 10 (0x0000000a)
```

However, it still seems to me that the debugging itself is not working properly:

```
(lldb) b -s helloworld.exe -a 0x140001338
Breakpoint 1: where = helloworld.exe`helloworld.exe[0x0000000140001338], address = 0x0000000140001338
(lldb) r
Process 9304 launched: 'C:\Users\xusheng\debugger_build\debugger\test\binaries\Windows-x86_64\helexe' (x86_64)
Process 9304 stopped
* thread #1, stop reason = breakpoint 1.1
    frame #0: 0x00007ff631a61338 helloworld.exe
->  0x7ff631a61338: int3
    0x7ff631a61339: subl   $0x28, %esp
    0x7ff631a6133c: callq  0x7ff631a6159c
    0x7ff631a61341: addq   $0x28, %rsp
```

For a helloworld.exe we use in our unit test, the LLDB when the entry breakpoint is hit, LLDB reports the byte as int3 instead of the original "sub rsp, 0x28". This issue is known in https://github.com/llvm/llvm-project/issues/60063, which actually prompted us to disable the LLDB adapter for PE files. In other words, LLDB 18.1 does get better in terms of remote debugging, but the debugging itself still malfunctions.

@25pwn I remember you said you can debug it normally -- are you not seeing the erroneous "int3" instruction while using the above setup?

**Reply by xusheng6** [reply]
However, it seems that the bug only happens when I set the breakpoint via `b -s -a` command. If I used `b symbol` to set a breakpoint on a function, it actually works:

```
(lldb) b __p___argv
Breakpoint 1: no locations (pending).
WARNING:  Unable to resolve breakpoint to any actual locations.
(lldb) r
1 location added to breakpoint 1
(lldb) Process 16572 launched: 'C:\Users\xusheng\debugger_build\debugger\test\binaries\Windows-x86_64\helloworld.exe' (x86_64)
Process 16572 stopped
* thread #1, stop reason = breakpoint 1.1
    frame #0: 0x00007ffe58372310 ucrtbase.dll`__p___argv
ucrtbase.dll`__p___argv:
->  0x7ffe58372310 <+0>:  leaq   0xfea81(%rip), %rax
    0x7ffe58372317 <+7>:  retq
    0x7ffe58372318 <+8>:  jno    0x7ffe5837236d            ; __stdio_common_vfwprintf + 61
    0x7ffe5837231a <+10>: popq   %rbp
(lldb) bt
* thread #1, stop reason = breakpoint 1.1
  * frame #0: 0x00007ffe58372310 ucrtbase.dll`__p___argv
    frame #1: 0x00007ff631a612b3 helloworld.exe
    frame #2: 0x00007ffe592c257d kernel32.dll`BaseThreadInitThunk + 29
    frame #3: 0x00007ffe5a9caf28 ntdll.dll`RtlUserThreadStart + 40
(lldb) breakpoint list
Current breakpoints:
1: name = '__p___argv', locations = 1, resolved = 1, hit count = 1
```

**Reply by xusheng6** [reply]
I have created an issue to track further development on the matter: https://github.com/Vector35/debugger/issues/613

**Reply by 25pwn** [reply]
I'd like to request re-enabling the lldb adapter for all platforms on dev(also stable if possible). The feature was definitely usable, and even if it's broken for some users, I think it's better to enable it for users that can make it work(maybe add a warning that it's experimental, or add it as an experimental option).

**Comment by xorhex** [comment]
Experiencing the same thing.  PE file is loaded into BN, but the connect to remote dbg server shows nothing in the drop down.

![image](https://github.com/user-attachments/assets/bf913b06-52bb-4734-9299-914458fbaf5e)

BN: 4.1.5902 (f2165c5d)
Host OS: Ubuntu 24.04.1 LTS

**Reply by xusheng6** [reply]
Remote debugging Windows PE from Linux is not supported yet. You can only do so from another Windows machine. https://github.com/Vector35/debugger/issues/70

**Comment by xusheng6** [comment]
I will continue to explore the possible route to support remote Windows debugging from a non-Windows OS, but it is not supported yet
